Домашнее задание 9 спринта, версия 1

DaniilKosmachev opened the pull request:
Домашнее задание спринта 9, версия 1

DaniilKosmachev committed:
Домашнее задание 9 спринта, версия 1

faritowich commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -13,9 +13,32 @@ class SettingsActivity : AppCompatActivity() {
+            shareIntent.putExtra(Intent.EXTRA_EMAIL, arrayOf(getString(R.string.email_support)))
+            shareIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.theme_support))
+            shareIntent.putExtra(Intent.EXTRA_TEXT,getString(R.string.message_for_support))
+            startActivity(shareIntent)
```
Можно ещё использовать scope функции, чтобы такие фрагменты кода сокращать до:

```
Intent(Intent.ACTION_SENDTO).apply {
                data= Uri.parse(getString(R.string.share_data_mailto))
                putExtra(Intent.EXTRA_EMAIL, arrayOf(getString(R.string.email_support)))
                putExtra(Intent.EXTRA_SUBJECT, getString(R.string.theme_support))
                putExtra(Intent.EXTRA_TEXT,getString(R.string.message_for_support))
                startActivity(this)
            }
```
Про такие методы почитать можно здесь:
https://kotlinlang.ru/docs/reference/scope-functions.html

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,11 +1,69 @@
 class SearchActivity : AppCompatActivity() {
+    companion object {
+        var searchQueryText = ""
+        const val SEARCH_STRING = "SEARCH_STRING"
```
Если планируем использовать только в данном классе, стоит сделать private

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,11 +1,69 @@
 
 class SearchActivity : AppCompatActivity() {
+    companion object {
+        var searchQueryText = ""
```
В companion object принято выносить константы, а переменные для конкретного экземпляра активити надо объявлять в самом классе.
Это связано с тем, что searchQueryText логически связано с конкретным экземпляром активити. А экземпляров класса в теории может быть сколько угодно, а сейчас searchQueryText привязан не к конкретному экземпляру класса, а к классу в целом. Константа же обычно не привязана к конкретному экземпляру, так что их вынесят в companion object - так константа будет одна для всех экземпляров класса, даже если их будет сто. А переменная, хранящая введенный текст, вообще не факт что постоянная для всех экземпляров класса - допустим, в разных экземплярах будет храниться разный текст, а если засунуть переменную в companion object, оно работать не будет. 
На самом деле в случае активити этот момент не является критичным, так как у нас в норме используется один экземпляр активити и разницы нет, но лучше иметь в виду этот момент и сразу делать так. То, что постоянно все зависимости от экземпляра класса - в константу в companion object, иначе - просто переменная.

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,11 +1,69 @@
+        const val SEARCH_STRING = "SEARCH_STRING"
+    }
+    override fun onSaveInstanceState(outState: Bundle) {
+        val editTextSearch = findViewById<EditText>(R.id.editTextSearchActivity)
```
1) Переменная объявляется и не используется
2) Сейчас для вью R.id.editTextSearchActivity переменная объявляется трижды в разных местах. Этого можно избежать, если объявлять глобальную переменную в теле класса с lateinit (будет private lateinit var viewName: EditText) или как нуллабл переменную (будет private var viewName: EditText? = null, (используется реже)), а в onCreate уже задавать значение findViewById<EditText>(R.id.VIEW_ID). Все остальные переменные, отвечающие за вьюхи, можно объявить так же. Тогда переменную можно инициализировать один раз в onCreate и иметь к ней доступ из любой функции класса.

DaniilKosmachev replied:
Вооот, lateinit, пробовал объявить как глобальную переменную (без lateinit)сразу инициализирована ее, но при переходе на активите, оно крашилось, потом соответственно понял, что я пытаюсь найти вьюху, пока ещё активите не создалось.
Сделаю

faritowich commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,91 @@
+        <item name="android:layout_width">match_parent</item>
+        <item name="android:layout_height">@dimen/main_button_height</item>
+        <item name="android:backgroundTint">@color/white</item>
+        <item name="android:layout_weight">1</item>
```
Лучше внешние отступы, высоту и ширину, веса, гравити не выносить в стиль. Если на экране будет много элементов, будет сложнее ориентироваться с их взаимным расположением, когда отвечающие за это атрибуты разнесены по разным файлам - что-то может быть в лейауте, что-то в стиле. В стилях обычно содержат атрибуты, отвечающие за внешний вид вьюхи и вообще только за то, что находится внутри нее - цвета, шрифты, фоны и т.д. Бывают исключения, когда, например, мы точно знаем, что у кнопки всегда должна строго определенная высота и ширина - тогда да, имеет смысл выносить

DaniilKosmachev replied:
Понял, в будущем учту

DaniilKosmachev committed:
Домашнее задание 9 спринта, версия 2
1) Изменены Intent на scope функции в SettingsActivity.kt
2) Константная переменная SEARCH_STRING сделана private
3) Переменная searchQueryText перенесена из companion object в глобаль
4) Все переменные вьюшек на SearchActivity.kt, SettingsActivity.kt и MainActivity.kt перенесены в глобаль и сделаны private lateinit var
